### ISSUE TITLE
Incorrect beaming when mixing notes and chords and stems are down

### ISSUE THREAD

**Issue by arshiacont** [grame-cncm/guidolib#161]
Issue when Stems are down only (up is fine).

Example:
```
{[ \clef<"g2"> \stemsUp \beamBegin:1 {g1/16, e2/16  } b0/16 e1/16 {b1/16, g2/16  } \beamEnd:1 
   \stemsDown \beamBegin:1 {b1/16, g2/16  } e1/16 g1/16 {e2/16, b2/16  } \beamEnd:1 
  ]}
```

![image](https://user-images.githubusercontent.com/3298204/185952512-3b7098cf-821f-498f-9615-92e0b6d5e926.png)

**Comment by arshiacont** [comment]
@dfober I believe I have a fix for this (and starting to understand the logic in GRBeams). Will PR soon.

In short: It seems like the Polygon points in GRBeam are set only using the first and last events in a Beam. In the above example, the _middle_ events should affect the position of the first point of the polygon (which by itself will lead to new stem lengths etc).

**Comment by dfober** [comment]
I have some doubts about this hypothesis : if you replace the last chord `{e2/16, b2/16  }` with a single note (e2 or b2), the beaming is correct. At the minimum it can be said that the chords interfere with the beaming in a different way.

**Comment by arshiacont** [comment]
Interesting.. in my fix, once the p0.y is set, I browse associated notes, look at their Stem Start Position and decide whether the initial decision was good or not (based on Direction). This works... but we might just be looking over a bug related to `getStemStartPos()` (which is the base decision for p0.y).

**Comment by arshiacont** [comment]
am looking at #154 right now which is somehow related to p0 initialisation.. will check this hypothesis

**Comment by dfober** [comment]
as far as I remember and in case of chords, a shared stem is used (GRGlobalStem)

**Comment by arshiacont** [comment]
You are absolutely right! my initial solution solves this special case but creates regressions elsewhere. Investigating further... .

**Comment by arshiacont** [comment]
@dfober I believe I have found the cause:

Chords seem to send `isEmpty()` as `true` which messes up with `GRBeam::setStemEndPos()` that fixes positions.

Is that a bug or is there a reason for this?

This explains why removing the chord at the end makes it work.

**Comment by arshiacont** [comment]
I confirm, after testing that the correct solution is to find a suitable way of avoiding `empty == true` for chords in `GRBeam::setStemEndPos()`. 

I can see elsewhere that this is also mentioned in the code (*). How would you suggest avoiding Empty for chords in the context?

(*) https://github.com/grame-cncm/guidolib/blob/bb00ffec5d84606c501a171fa5b6183a8e71fbb5/src/engine/graphic/GRSlur.cpp#L144

**Comment by arshiacont** [comment]
Or maybe simply the current definition of isEmpty used in `GRBeam::setStemEndPos()` is wrong:
https://github.com/grame-cncm/guidolib/blob/6d1fb4474af0b95dde8668188db904c0baba2074/src/engine/graphic/GRBeam.cpp#L869 
instead of `bool empty = (sn->isEmpty() && !sn->getDuration());` ?

Once again: I confirm that this potentially closes the issue.

**Comment by dfober** [comment]
Chord encoding in guido is very special. A chord does not really exist as an object, it is represented in a similar way to gmn syntax, framed by `empty`, the first one carries the duration of the chord and the last one closes the chord with a zero duration. For example:
~~~~~~~~
{ c, e, g } 
~~~~~~~~
will be represented in AR by 
~~~~~~~~
\chordTag \shareLocation \displayDuration \shareStem 
empty*1/4 \chordComma c1*0/1 \chordComma e1*0/1 \chordComma g1*0/1 empty*0/1 
~~~~~~~~
This is why it's necessary to distinguish between 'real' empty and those which frame chords.

Regarding the proposed change with the empty computation, I'll have a look as soon as I'll be back home. Did you checked that it works in any case with  the validation tests ?

**Comment by dfober** [comment]
fixed.

> Or maybe simply the current definition of isEmpty used in `GRBeam::setStemEndPos()` is wrong:
> 
> https://github.com/grame-cncm/guidolib/blob/6d1fb4474af0b95dde8668188db904c0baba2074/src/engine/graphic/GRBeam.cpp#L869
> 
> 
> instead of `bool empty = (sn->isEmpty() && !sn->getDuration());` ?

Was a bit more complex than just changing the test (this was breaking the solved issue #146)
